lib/domain/validation/authenticator_validator.dart

```diff
@@ -0,0 +1,116 @@
+// Copyright © 2024 Nevis Security AG. All rights reserved.
+
+import 'dart:io' show Platform;
+
+import 'package:injectable/injectable.dart';
+import 'package:nevis_mobile_authentication_sdk/nevis_mobile_authentication_sdk.dart';
+import 'package:nevis_mobile_authentication_sdk_example_app_flutter/util/collection_utils.dart';
+
+abstract class AuthenticatorValidator {
+  Future<Set<Authenticator>> validateForRegistration(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  );
+
+  Set<Authenticator> validateForAuthentication(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  );
+}
+
+@Injectable(as: AuthenticatorValidator)
+class AuthenticatorValidatorImpl implements AuthenticatorValidator {
+  @override
+  Future<Set<Authenticator>> validateForRegistration(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    Set<Authenticator> allowedAuthenticators = _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    );
+
+    return whereAsync(allowedAuthenticators, (authenticator) async {
+      // Do not display:
+      //  - policy non-compliant authenticators (this includes already registered authenticators)
+      //  - not hardware supported authenticators
+      //  - prefer Biometrics authenticator on Android
+      return authenticator.isSupportedByHardware &&
+          await context.isPolicyCompliant(authenticator.aaid) &&
+          await _filterAndroidFingerprintIfNecessary(context, authenticator);
+    });
+  }
+
+  @override
+  Set<Authenticator> validateForAuthentication(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    return _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    ).where((authenticator) {
+      // Do not display:
+      //   - non-registered authenticators
+      //   - not hardware supported authenticators
+      return authenticator.registration
+              .isRegistered(context.account.username) &&
+          authenticator.isSupportedByHardware;
+    }).toSet();
+  }
```

Comment:
❓Do I see correctly that policy compliance is not checked when filtering/validating authenticators for authentication? And if so, is this right? We do check this in the Android Access App at least: https://github.com/nevissecurity/nevis-access-app-android/blob/b8bd27158e15641e12ce02101a8a0c1b6462ff75/showaccess-app/src/main/kotlin/ch/nevis/accessapp/util/AuthenticationAuthenticatorFilter.kt

Reply:
In case of all example apps non-policy compliant authenticators are displayed as well but not enabled for selection.

Reply:
Okay.